### PR TITLE
Add logic to update configMaps on operator

### PR DESCRIPTION
**Pull request by khaosdoctor**
Adds the logic to save the routing tables in the config maps

### CODE REVIEW

File: operator/controllers/routing_table.go [hunk 1]
```diff
@@ -58,12 +54,44 @@ func addAndUpdateRoutingTable(
...
+) error {
+	tableAsJSON, marshalErr := table.MarshalJSON()
+	if marshalErr != nil { // should never happen
+		return marshalErr
+	}
+
+	routingConfigMap, err := k8s.GetConfigMap(ctx, cl, namespace, routingTableName)
+	if err != nil { return err }
+
+	if routingConfigMap == nil { // if the routing table doesn't exist, we need to create it with the latest data
```
Comment (arschles) [inline operator/controllers/routing_table.go]: @khaosdoctor can you put the comment above this line?

```suggestion
        // if the routing table doesn't exist, we need to create it with the latest data
	if routingConfigMap == nil {
```

File: operator/controllers/routing_table.go [hunk 2]
```diff
@@ -58,12 +54,44 @@ func addAndUpdateRoutingTable(
...
+	table *routing.Table,
+	lggr logr.Logger,
+) error {
+	tableAsJSON, marshalErr := table.MarshalJSON()
+	if marshalErr != nil { // should never happen
+		return marshalErr
+	}
+
+	routingConfigMap, err := k8s.GetConfigMap(ctx, cl, namespace, routingTableName)
+	if err != nil { return err }
```
Comment (arschles) [inline operator/controllers/routing_table.go]:
```suggestion
	if err != nil {
	    return err
	}
```

File: operator/controllers/routing_table.go [hunk 3]
```diff
@@ -58,12 +54,44 @@ func addAndUpdateRoutingTable(
...
+
+func updateRoutingMap(
+	ctx context.Context,
+	cl client.Client,
+	namespace string,
+	table *routing.Table,
+	lggr logr.Logger,
+) error {
+	tableAsJSON, marshalErr := table.MarshalJSON()
+	if marshalErr != nil { // should never happen
```
Comment (arschles) [inline operator/controllers/routing_table.go]:
```suggestion
	if marshalErr != nil {
```

File: operator/controllers/routing_table.go [hunk 4]
```diff
@@ -58,12 +54,44 @@ func addAndUpdateRoutingTable(
...
-	)
+	return updateRoutingMap(ctx, cl, namespace, table, lggr)
+}
+
+func updateRoutingMap(
+	ctx context.Context,
+	cl client.Client,
+	namespace string,
+	table *routing.Table,
+	lggr logr.Logger,
```
Comment (arschles) [inline operator/controllers/routing_table.go]: let's put the `lggr` parameter after `ctx`, to be consistent

```suggestion
	ctx context.Context,
	lggr logr.Logger,
	cl client.Client,
	namespace string,
	table *routing.Table,
```

File: operator/controllers/routing_table.go [hunk 5]
```diff
@@ -58,12 +54,44 @@ func addAndUpdateRoutingTable(
...
+			"name": "http-add-on-routing-table",
+		}
+
+		if err := k8s.CreateConfigMap(ctx, cl, k8s.NewConfigMap(namespace, routingTableName, routingTableLabels, routingTableData), lggr); err != nil {
+			return err
+		}
+	} else {
+		newRoutingTable := routingConfigMap.DeepCopy()
+		newRoutingTable.Data["routing-table"] = string(tableAsJSON)
+		if _, patchErr := k8s.PatchConfigMap(ctx, cl, lggr, routingConfigMap, newRoutingTable); patchErr != nil { return patchErr }
```
Comment (arschles) [inline operator/controllers/routing_table.go]:
```suggestion
		if _, patchErr := k8s.PatchConfigMap(ctx, cl, lggr, routingConfigMap, newRoutingTable); patchErr != nil {
		    return patchErr
		}
```

File: pkg/k8s/configMap.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	}
+
+	return configMap
+}
+
+func CreateConfigMap (
+	ctx context.Context,
+	cl client.Client,
+	configMap *corev1.ConfigMap,
+	logger logr.Logger,
```
Comment (arschles) [inline pkg/k8s/configMap.go]: let's put the logger next to `ctx` to be consistent

```suggestion
	ctx context.Context,
	logger logr.Logger,
	cl client.Client,
	configMap *corev1.ConfigMap,
```

File: pkg/k8s/configMap.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+) error {
+	 existentConfigMap, err := GetConfigMap(ctx, cl, configMap.Namespace, configMap.Name)
+	if err == nil { return err }
```
Comment (arschles) [inline pkg/k8s/configMap.go]:
```suggestion
	if err == nil {
	    return err
	}
```

File: pkg/k8s/configMap.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+	if existentConfigMap.Name != "" { return errors.New("ConfigMap already exists") }
```
Comment (arschles) [inline pkg/k8s/configMap.go]:
```suggestion
	if existentConfigMap.Name != "" {
	    return fmt.Errorf("ConfigMap %s already exists", configMap.Name)
	}
```

File: pkg/k8s/configMap.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+
+	createErr := cl.Create(ctx, configMap)
+	if createErr != nil { return createErr }
```
Comment (arschles) [inline pkg/k8s/configMap.go]:
```suggestion
	if createErr != nil {
	    return createErr
	}
```

File: pkg/k8s/configMap.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+func DeleteConfigMap (
+	ctx context.Context,
+	cl client.Client,
+	configMap *corev1.ConfigMap,
+	logger logr.Logger,
+) error {
+	err := cl.Delete(ctx, configMap)
+	if err != nil { return err }
+	return nil
```
Comment (arschles) [inline pkg/k8s/configMap.go]:
```suggestion
	return cl.Delete(ctx, configMap)
```

File: pkg/k8s/configMap.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+	return patchConfigMap, nil
+}
+
+func GetConfigMap (
+	ctx context.Context,
+	cl client.Client,
+	namespace string,
+	name string,
+) (*corev1.ConfigMap, error) {
+
```
Comment (arschles) [inline pkg/k8s/configMap.go]:
```suggestion
```

**Review by arschles — CHANGES_REQUESTED** [review]
@khaosdoctor left some suggestions. this is a good start. let's get this merged and then I'll go and implement the interceptor side of things.

**Comment by arschles** [comment]
@khaosdoctor I think you also will need to delete the "admin" server endpoints in the operator

**Comment by khaosdoctor** [comment]
@arschles Guess I removed everything. Can you check?

**Pull request merged**